In [1]:
import pandas as pd
import numpy as np

# PM2.5 from 2013 to 2022

## PM2.5 in 2013

In [49]:

# Read the 2013 PM2.5 file, skipping the first 7 lines of metadata
PM25_2013 = pd.read_csv(
    r"air_pollution_data\PM25_2013.csv",
    skiprows=7,      # Skip lines 1–7 (6 description lines + 1 blank line)
    header=0,        # Use line 8 as the column names
    nrows=79570,     # Read the remaining 79,570 data rows (from line 9 to line 79,578)
    engine="python"  # Ensure skiprows + nrows are handled properly
)

# Keep only the English part of each column name (text before "//") and strip whitespace
PM25_2013.columns = [col.split("//")[0].strip() for col in PM25_2013.columns]

#Filter for Ontario
ontario_PM25_2013 = PM25_2013[PM25_2013["Province/Territory"] == "ON"].copy()

# Verify the cleaned column names and dataset shape
print(PM25_2013.columns.tolist())

print(f"All rows: {PM25_2013.shape[0]}, Ontario rows: {ontario_PM25_2013.shape[0]}")

                                                                    
                   

['Pollutant', 'Method Code', 'NAPS ID', 'City', 'Province/Territory', 'Latitude', 'Longitude', 'Date', 'H01', 'H02', 'H03', 'H04', 'H05', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24']
All rows: 79570, Ontario rows: 15695


In [93]:
PM25_2013.head(6)

Pollutant  Method Code  NAPS ID    City Province/Territory  Latitude  \
0     PM2.5          184    60104  Ottawa                 ON  45.43433   
1     PM2.5          184    60104  Ottawa                 ON  45.43433   
2     PM2.5          184    60104  Ottawa                 ON  45.43433   
3     PM2.5          184    60104  Ottawa                 ON  45.43433   
4     PM2.5          184    60104  Ottawa                 ON  45.43433   
5     PM2.5          184    60104  Ottawa                 ON  45.43433   

   Longitude        Date  H01  H02  ...  H15  H16  H17  H18  H19  H20  H21  \
0    -75.676  2013-01-01    2    2  ...    1    2    3    4    8    9   17   
1    -75.676  2013-01-02   13    8  ...    2    2    3    8   11   15   22   
2    -75.676  2013-01-03   34   29  ... -999 -999 -999 -999 -999 -999 -999   
3    -75.676  2013-01-04 -999 -999  ...    9    9    8    8    7    7    5   
4    -75.676  2013-01-05    3    1  ...    2    2    3   10   17   14   32   
5    -75.676  2013-01-06   12   11  ...   16   14   16   18   19   22    6   

   H22  H23  H24  
0   12   10   10  
1   32   40   34  
2 -999 -999 -999  
3    4    4    4  
4   28   26   18  
5    3    3    3  

[6 rows x 32 columns]

In [94]:
PM25_2013.tail(6)

Pollutant  Method Code  NAPS ID        City Province/Territory  \
79564     PM2.5          195    90606  Bruderheim                 AB   
79565     PM2.5          195    90606  Bruderheim                 AB   
79566     PM2.5          195    90606  Bruderheim                 AB   
79567     PM2.5          195    90606  Bruderheim                 AB   
79568     PM2.5          195    90606  Bruderheim                 AB   
79569     PM2.5          195    90606  Bruderheim                 AB   

       Latitude  Longitude        Date  H01  H02  ...  H15  H16  H17  H18  \
79564  53.79999 -112.92781  2013-12-26    2    2  ...    1    1    2    2   
79565  53.79999 -112.92781  2013-12-27    0    1  ...    3    3    2    3   
79566  53.79999 -112.92781  2013-12-28    4    4  ...    8    9   15   13   
79567  53.79999 -112.92781  2013-12-29   14    9  ...    9    9    8    7   
79568  53.79999 -112.92781  2013-12-30    8    7  ...    5    4    5    4   
79569  53.79999 -112.92781  2013-12-31    4    5  ...    4    5    6    6   

       H19  H20  H21  H22  H23  H24  
79564    4    1    1    1    0    0  
79565    3    6    3    2    3    4  
79566   26   19   25   16   10    7  
79567    7    7    7    6    7    7  
79568    4    3    3    3    3    4  
79569    7   10   14   13   10    9  

[6 rows x 32 columns]

In [50]:
# Display the first 6 rows in a DataFrame format
ontario_PM25_2013.head(6)


Pollutant  Method Code  NAPS ID    City Province/Territory  Latitude  \
0     PM2.5          184    60104  Ottawa                 ON  45.43433   
1     PM2.5          184    60104  Ottawa                 ON  45.43433   
2     PM2.5          184    60104  Ottawa                 ON  45.43433   
3     PM2.5          184    60104  Ottawa                 ON  45.43433   
4     PM2.5          184    60104  Ottawa                 ON  45.43433   
5     PM2.5          184    60104  Ottawa                 ON  45.43433   

   Longitude        Date  H01  H02  ...  H15  H16  H17  H18  H19  H20  H21  \
0    -75.676  2013-01-01    2    2  ...    1    2    3    4    8    9   17   
1    -75.676  2013-01-02   13    8  ...    2    2    3    8   11   15   22   
2    -75.676  2013-01-03   34   29  ... -999 -999 -999 -999 -999 -999 -999   
3    -75.676  2013-01-04 -999 -999  ...    9    9    8    8    7    7    5   
4    -75.676  2013-01-05    3    1  ...    2    2    3   10   17   14   32   
5    -75.676  2013-01-06   12   11  ...   16   14   16   18   19   22    6   

   H22  H23  H24  
0   12   10   10  
1   32   40   34  
2 -999 -999 -999  
3    4    4    4  
4   28   26   18  
5    3    3    3  

[6 rows x 32 columns]

In [51]:
#Display info
ontario_PM25_2013.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15695 entries, 0 to 15694
Data columns (total 32 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Pollutant           15695 non-null  object 
 1   Method Code         15695 non-null  int64  
 2   NAPS ID             15695 non-null  int64  
 3   City                15695 non-null  object 
 4   Province/Territory  15695 non-null  object 
 5   Latitude            15695 non-null  float64
 6   Longitude           15695 non-null  float64
 7   Date                15695 non-null  object 
 8   H01                 15695 non-null  int64  
 9   H02                 15695 non-null  int64  
 10  H03                 15695 non-null  int64  
 11  H04                 15695 non-null  int64  
 12  H05                 15695 non-null  int64  
 13  H06                 15695 non-null  int64  
 14  H07                 15695 non-null  int64  
 15  H08                 15695 non-null  int64  
 16  H09      

In [52]:
# Unique values in City

# 1. Number of unique cities
num_cities = ontario_PM25_2013["City"].nunique(dropna=True)
print(f"Total unique cities: {num_cities}")

# 2. List of unique city names
unique_cities = ontario_PM25_2013["City"].dropna().unique()
print("Unique city names:")
print(unique_cities)

# 3. How many readings per city
city_counts = ontario_PM25_2013["City"].value_counts()
print("\nObservations per city:")
print(city_counts)


Total unique cities: 34
Unique city names:
['Ottawa' 'Windsor' 'Kingston' 'Toronto' 'Brampton' 'Mississauga'
 'Hamilton' 'Sudbury' 'Sault Ste. Marie' 'Thunder Bay' 'London' 'Sarnia'
 'Peterborough' 'Cornwall' 'St. Catharines' 'Brantford' 'Kitchener'
 'Oakville' 'Oshawa' 'Guelph' 'North Bay' 'Tiverton' 'Simcoe' 'Burlington'
 'Dorset' 'Grand Bend' 'Barrie' 'Newmarket' 'Parry Sound' 'Port Stanley'
 'Belleville' 'Morrisburg' 'Chatham' 'Petawawa']

Observations per city:
City
Toronto             1825
Hamilton            1095
Windsor              730
Ottawa               730
Kingston             730
Brampton             365
Mississauga          365
Sudbury              365
Sault Ste. Marie     365
Thunder Bay          365
London               365
Sarnia               365
Peterborough         365
Cornwall             365
St. Catharines       365
Brantford            365
Kitchener            365
Oakville             365
Oshawa               365
Guelph               365
North Bay            365

From the results we can see that there are duplicates in some cities, like Toronto, Hamilton, Windsor

In [53]:
# "NAPS ID" means Station
station_counts = ontario_PM25_2013.groupby("City")["NAPS ID"].nunique()
print(station_counts)


City
Barrie              1
Belleville          1
Brampton            1
Brantford           1
Burlington          1
Chatham             1
Cornwall            1
Dorset              1
Grand Bend          1
Guelph              1
Hamilton            3
Kingston            2
Kitchener           1
London              1
Mississauga         1
Morrisburg          1
Newmarket           1
North Bay           1
Oakville            1
Oshawa              1
Ottawa              2
Parry Sound         1
Petawawa            1
Peterborough        1
Port Stanley        1
Sarnia              1
Sault Ste. Marie    1
Simcoe              1
St. Catharines      1
Sudbury             1
Thunder Bay         1
Tiverton            1
Toronto             5
Windsor             2
Name: NAPS ID, dtype: int64


From the results, we can confirm that the duplicates come from mutiple stations. For example, in Toronto, there are 5 stations, so Totonto has 1825 values in 2013

### Data Type

In [54]:
#Convert 'Data' from object to Datetime

ontario_PM25_2013["Date"] = pd.to_datetime(ontario_PM25_2013["Date"], format="%Y-%m-%d", errors="raise")
print(ontario_PM25_2013.info())

<class 'pandas.core.frame.DataFrame'>
Index: 15695 entries, 0 to 15694
Data columns (total 32 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           15695 non-null  object        
 1   Method Code         15695 non-null  int64         
 2   NAPS ID             15695 non-null  int64         
 3   City                15695 non-null  object        
 4   Province/Territory  15695 non-null  object        
 5   Latitude            15695 non-null  float64       
 6   Longitude           15695 non-null  float64       
 7   Date                15695 non-null  datetime64[ns]
 8   H01                 15695 non-null  int64         
 9   H02                 15695 non-null  int64         
 10  H03                 15695 non-null  int64         
 11  H04                 15695 non-null  int64         
 12  H05                 15695 non-null  int64         
 13  H06                 15695 non-null  int64         


### Calculate 3h Max for each day and each station

3hMax (short for 3-hour maximum) is a common metric used in air quality and environmental monitoring. It refers to the maximum average concentration of a pollutant measured over any consecutive 3-hour period within a day.

In [55]:
# 1. Define the hourly columns H01 through H24
hour_cols = [f'H{i:02d}' for i in range(1, 25)]

# 2. Replace invalid values (-999) with NaN so they are excluded from calculations
ontario_PM25_2013[hour_cols] = ontario_PM25_2013[hour_cols].replace({-999: np.nan})

# 3. Compute the 3-hour rolling average for each row, then take the maximum of those averages
ontario_PM25_2013['PM25_3hMax'] = (
    ontario_PM25_2013[hour_cols]
    .apply(lambda row: row.rolling(window=3, min_periods=3).mean().max(), axis=1)
).round(2)

# 4. Print the first few rows to verify the new column
print(ontario_PM25_2013[['Date','PM25_3hMax']].head())


        Date  PM25_3hMax
0 2013-01-01       13.00
1 2013-01-02       35.33
2 2013-01-03       28.33
3 2013-01-04       10.00
4 2013-01-05       28.67


In [56]:
# Count each unique value and show the result
counts_3hMax = ontario_PM25_2013["PM25_3hMax"].value_counts().sort_index()

# Display the result
print(counts_3hMax)

PM25_3hMax
0.00      8
0.33      4
0.67      7
1.00     31
1.33     19
         ..
73.00     1
75.33     1
82.00     1
92.00     1
96.33     1
Name: count, Length: 182, dtype: int64


In [57]:
ontario_PM25_2013.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15695 entries, 0 to 15694
Data columns (total 33 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           15695 non-null  object        
 1   Method Code         15695 non-null  int64         
 2   NAPS ID             15695 non-null  int64         
 3   City                15695 non-null  object        
 4   Province/Territory  15695 non-null  object        
 5   Latitude            15695 non-null  float64       
 6   Longitude           15695 non-null  float64       
 7   Date                15695 non-null  datetime64[ns]
 8   H01                 14962 non-null  float64       
 9   H02                 14961 non-null  float64       
 10  H03                 14968 non-null  float64       
 11  H04                 14971 non-null  float64       
 12  H05                 14972 non-null  float64       
 13  H06                 14959 non-null  float64       


### Calculate 8h Max

In [58]:

# 1. Define the hourly columns H01 through H24
hour_cols = [f'H{i:02d}' for i in range(1, 25)]

# 2. Replace invalid values (-999) with NaN so they are excluded from calculations
ontario_PM25_2013[hour_cols] = ontario_PM25_2013[hour_cols].replace({-999: np.nan})

# 3. Compute the 8-hour rolling average for each row, then take the maximum of those averages
ontario_PM25_2013['PM25_8hMax'] = (
    ontario_PM25_2013[hour_cols]
    .apply(lambda row: row.rolling(window=8, min_periods=8).mean().max(), axis=1)
).round(2)

# 4. Print the first few rows to verify the new column
print(ontario_PM25_2013[['Date','PM25_8hMax']].head())


        Date  PM25_8hMax
0 2013-01-01        9.12
1 2013-01-02       20.62
2 2013-01-03       22.62
3 2013-01-04        9.25
4 2013-01-05       18.50


In [59]:
# Count each unique value and show the result
counts_8hMax = ontario_PM25_2013["PM25_8hMax"].value_counts().sort_index()

# Display the result
print(counts_8hMax)

PM25_8hMax
0.00     8
0.12     4
0.25     4
0.38     4
0.50     4
        ..
54.38    2
55.50    2
62.25    1
67.00    1
78.00    1
Name: count, Length: 356, dtype: int64


In [60]:
ontario_PM25_2013.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15695 entries, 0 to 15694
Data columns (total 34 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           15695 non-null  object        
 1   Method Code         15695 non-null  int64         
 2   NAPS ID             15695 non-null  int64         
 3   City                15695 non-null  object        
 4   Province/Territory  15695 non-null  object        
 5   Latitude            15695 non-null  float64       
 6   Longitude           15695 non-null  float64       
 7   Date                15695 non-null  datetime64[ns]
 8   H01                 14962 non-null  float64       
 9   H02                 14961 non-null  float64       
 10  H03                 14968 non-null  float64       
 11  H04                 14971 non-null  float64       
 12  H05                 14972 non-null  float64       
 13  H06                 14959 non-null  float64       


### Annual Max 3-hour/8-hour Average PM2.5 Concentration for each Station (NAPS ID )/City

Since we want to calculate annual mean, so we don't need to fill the missing values for 3hMax and 8hMax, unless no data in a station

In [84]:


# 1. Extract the year from the Date column
ontario_PM25_2013['Year'] = pd.to_datetime(ontario_PM25_2013['Date']).dt.year

# 2. Group by Year, City, and NAPS ID to compute annual mean, 
#    and carry through station metadata
PM25_annual_mean_2013 = (
    ontario_PM25_2013
    .groupby(['Year', 'City', 'NAPS ID'], as_index=False)
    .agg(
        Pollutant    = ('Pollutant',    'first'),
        Latitude     = ('Latitude',     'first'),
        Longitude    = ('Longitude',    'first'),
        PM25_annu_mean_3hMax = ('PM25_3hMax',       'mean'),           
        PM25_annu_mean_8hMax = ('PM25_8hMax',       'mean') )           
)

# 3. Sort for readability
PM25_annual_mean_2013.sort_values(['Year', 'City', 'NAPS ID'], inplace=True)

# 4. Inspect the first few rows
print(PM25_annual_mean_2013.head())

# 5. Check the structure of the result
PM25_annual_mean_2013.info()




   Year        City  NAPS ID Pollutant  Latitude  Longitude  \
0  2013      Barrie    65001     PM2.5  44.38231  -79.70243   
1  2013  Belleville    65401     PM2.5  44.15053  -77.39550   
2  2013    Brampton    60428     PM2.5  43.69875  -79.78092   
3  2013   Brantford    61402     PM2.5  43.13861  -80.29264   
4  2013  Burlington    63001     PM2.5  43.31457  -79.80276   

   PM25_annu_mean_3hMax  PM25_annu_mean_8hMax  
0             13.350914             10.700554  
1             12.234658              9.900630  
2             15.326978             11.972121  
3             14.327821             11.783567  
4             14.557967             12.110165  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43 entries, 0 to 42
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  43 non-null     int32  
 1   City                  43 non-null     object 
 2   NAPS ID               

Check missing values for annual_max_3h and annual_max_8h. The station '60304' in Kingston never produced any valid data in 2013, need to check other years to see how to cope this issue

In [64]:
# 1. Which stations got zero valid daily 3hMax values?
zero_3h = (
    ontario_PM25_2013
    .groupby('NAPS ID')['PM25_3hMax']
    .apply(lambda s: s.notna().sum())
)
print("Stations with no valid 3hMax days:")
print(zero_3h[zero_3h == 0])


# 2. Same for 8-hour
zero_8h = (
    ontario_PM25_2013
    .groupby('NAPS ID')['PM25_8hMax']
    .apply(lambda s: s.notna().sum())
)
print("\nStations with no valid 8hMax days:")
print(zero_8h[zero_8h == 0])

# 3. Inspect one of those stations’ daily data to see why
bad_id = zero_3h[zero_3h == 0].index[0]
print(f"\nDaily 3hMax for station {bad_id}:")
print(ontario_PM25_2013.loc[ontario_PM25_2013['NAPS ID']==bad_id, ['Date','PM25_3hMax']].head(10))


Stations with no valid 3hMax days:
NAPS ID
60304    0
Name: PM25_3hMax, dtype: int64

Stations with no valid 8hMax days:
NAPS ID
60304    0
Name: PM25_8hMax, dtype: int64

Daily 3hMax for station 60304:
           Date  PM25_3hMax
1825 2013-01-01         NaN
1826 2013-01-02         NaN
1827 2013-01-03         NaN
1828 2013-01-04         NaN
1829 2013-01-05         NaN
1830 2013-01-06         NaN
1831 2013-01-07         NaN
1832 2013-01-08         NaN
1833 2013-01-09         NaN
1834 2013-01-10         NaN


In [65]:
# 4. Save out to CSV 
out_path = r"air_pollution_data\PM25_2013_ontario_annual_by_station.csv"
PM25_annual_mean_2013.to_csv(out_path, index=False)
print(f"Saved annual station stats to {out_path}")

Saved annual station stats to air_pollution_data\PM25_2013_ontario_annual_by_station.csv


# NO2 from 2013_2022

## NO2 in 2013

In [66]:
# Read the 2013 NO2 file, skipping the first 7 lines of metadata
NO2_2013 = pd.read_csv(
    r"air_pollution_data\NO2_2013.csv",
    skiprows=7,      # Skip lines 1–7 (6 description lines + 1 blank line)
    header=0,        # Use line 8 as the column names
    nrows=60955,     # Read the remaining 60,955 data rows (from line 9 to line 60963)
    engine="python"  # Ensure skiprows + nrows are handled properly
)

# Keep only the English part of each column name (text before "//") and strip whitespace
NO2_2013.columns = [col.split("//")[0].strip() for col in NO2_2013.columns]

#Filter for Ontario
ontario_NO2_2013 = NO2_2013[NO2_2013["Province/Territory"] == "ON"].copy()

# Verify the cleaned column names and dataset shape
print(NO2_2013.columns.tolist())

print(f"All rows: {NO2_2013.shape[0]}, Ontario rows: {ontario_NO2_2013.shape[0]}")




['Pollutant', 'NAPS ID', 'City', 'Province/Territory', 'Latitude', 'Longitude', 'Date', 'H01', 'H02', 'H03', 'H04', 'H05', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24']
All rows: 60955, Ontario rows: 14235


In [95]:
NO2_2013.head(6)

Pollutant  NAPS ID        City Province/Territory  Latitude  Longitude  \
0       NO2    10102  St. John's                 NL  47.56038  -52.71147   
1       NO2    10102  St. John's                 NL  47.56038  -52.71147   
2       NO2    10102  St. John's                 NL  47.56038  -52.71147   
3       NO2    10102  St. John's                 NL  47.56038  -52.71147   
4       NO2    10102  St. John's                 NL  47.56038  -52.71147   
5       NO2    10102  St. John's                 NL  47.56038  -52.71147   

         Date  H01  H02  H03  ...  H15  H16  H17  H18  H19  H20  H21  H22  \
0  2013-01-01    2    2    1  ...    2    2    2    4    5    5    6    3   
1  2013-01-02    4    4    5  ...    4    6   13   10   13   13   13   14   
2  2013-01-03    3    2    3  ...    7    6    6    9   14    4    5    3   
3  2013-01-04    3    2    3  ...    6    7    8    6    6    5    4    6   
4  2013-01-05    5    5    5  ...    4    3    4    4    3    5    7    5   
5  2013-01-06    2    1    1  ...    2    5    6    6    6    4    4    4   

   H23  H24  
0    3    4  
1    7    5  
2    3    4  
3    7    5  
4    3    2  
5    3    3  

[6 rows x 31 columns]

In [96]:
NO2_2013.tail(6)

Pollutant  NAPS ID     City Province/Territory  Latitude  Longitude  \
60949       NO2   129303  Iqaluit                 NU  63.75162  -68.52243   
60950       NO2   129303  Iqaluit                 NU  63.75162  -68.52243   
60951       NO2   129303  Iqaluit                 NU  63.75162  -68.52243   
60952       NO2   129303  Iqaluit                 NU  63.75162  -68.52243   
60953       NO2   129303  Iqaluit                 NU  63.75162  -68.52243   
60954       NO2   129303  Iqaluit                 NU  63.75162  -68.52243   

             Date  H01  H02  H03  ...  H15  H16  H17  H18  H19  H20  H21  H22  \
60949  2013-12-26 -999 -999 -999  ... -999 -999 -999 -999 -999 -999 -999 -999   
60950  2013-12-27 -999 -999 -999  ... -999 -999 -999 -999 -999 -999 -999 -999   
60951  2013-12-28 -999 -999 -999  ... -999 -999 -999 -999 -999 -999 -999 -999   
60952  2013-12-29 -999 -999 -999  ... -999 -999 -999 -999 -999 -999 -999 -999   
60953  2013-12-30 -999 -999 -999  ... -999 -999 -999 -999 -999 -999 -999 -999   
60954  2013-12-31 -999 -999 -999  ... -999 -999 -999 -999 -999 -999 -999 -999   

       H23  H24  
60949 -999 -999  
60950 -999 -999  
60951 -999 -999  
60952 -999 -999  
60953 -999 -999  
60954 -999    0  

[6 rows x 31 columns]

In [67]:
# Display the first 6 rows in a DataFrame format
ontario_NO2_2013.head(6)

Pollutant  NAPS ID    City Province/Territory  Latitude  Longitude  \
13505       NO2    60104  Ottawa                 ON  45.43433    -75.676   
13506       NO2    60104  Ottawa                 ON  45.43433    -75.676   
13507       NO2    60104  Ottawa                 ON  45.43433    -75.676   
13508       NO2    60104  Ottawa                 ON  45.43433    -75.676   
13509       NO2    60104  Ottawa                 ON  45.43433    -75.676   
13510       NO2    60104  Ottawa                 ON  45.43433    -75.676   

             Date  H01  H02  H03  ...  H15  H16  H17  H18  H19  H20  H21  H22  \
13505  2013-01-01    3    3    3  ...    3    4    7   10   16   14   28   26   
13506  2013-01-02   40   33   28  ...    3    4    8   12   17   17   15   24   
13507  2013-01-03   12   11   15  ... -999 -999 -999 -999 -999 -999 -999 -999   
13508  2013-01-04 -999 -999 -999  ...   19   25   21   18   14   12   12   10   
13509  2013-01-05    5 -999 -999  ... -999 -999 -999 -999 -999 -999 -999 -999   
13510  2013-01-06 -999 -999 -999  ... -999 -999 -999 -999 -999 -999 -999 -999   

       H23  H24  
13505   25   32  
13506   21   17  
13507 -999 -999  
13508   10    9  
13509 -999 -999  
13510 -999 -999  

[6 rows x 31 columns]

In [68]:
#Display info
ontario_NO2_2013.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14235 entries, 13505 to 27739
Data columns (total 31 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Pollutant           14235 non-null  object 
 1   NAPS ID             14235 non-null  int64  
 2   City                14235 non-null  object 
 3   Province/Territory  14235 non-null  object 
 4   Latitude            14235 non-null  float64
 5   Longitude           14235 non-null  float64
 6   Date                14235 non-null  object 
 7   H01                 14235 non-null  int64  
 8   H02                 14235 non-null  int64  
 9   H03                 14235 non-null  int64  
 10  H04                 14235 non-null  int64  
 11  H05                 14235 non-null  int64  
 12  H06                 14235 non-null  int64  
 13  H07                 14235 non-null  int64  
 14  H08                 14235 non-null  int64  
 15  H09                 14235 non-null  int64  
 16  H10  

In [69]:
# Unique values in City

# 1. Number of unique cities
num_cities = ontario_NO2_2013["City"].nunique(dropna=True)
print(f"Total unique cities: {num_cities}")

# 2. List of unique city names
unique_cities = ontario_NO2_2013["City"].dropna().unique()
print("Unique city names:")
print(unique_cities)

# 3. How many readings per city
city_counts = ontario_NO2_2013["City"].value_counts()
print("\nObservations per city:")
print(city_counts)



Total unique cities: 30
Unique city names:
['Ottawa' 'Windsor' 'Kingston' 'Toronto' 'Brampton' 'Mississauga'
 'Hamilton' 'Sudbury' 'Sault Ste. Marie' 'Thunder Bay' 'London' 'Sarnia'
 'Peterborough' 'Cornwall' 'St. Catharines' 'Brantford' 'Kitchener'
 'Oakville' 'Oshawa' 'Guelph' 'North Bay' 'Tiverton' 'Simcoe' 'Burlington'
 'Grand Bend' 'Barrie' 'Newmarket' 'Parry Sound' 'Belleville' 'Chatham']

Observations per city:
City
Toronto             1825
Hamilton            1095
Windsor              730
Kingston             730
Ottawa               730
Brampton             365
Mississauga          365
Sudbury              365
Sault Ste. Marie     365
Thunder Bay          365
London               365
Sarnia               365
Peterborough         365
Cornwall             365
St. Catharines       365
Brantford            365
Kitchener            365
Oakville             365
Oshawa               365
Guelph               365
North Bay            365
Tiverton             365
Simcoe               36

From the results we can see that there are duplicates in some cities, like Toronto, Hamilton, Windsor


In [70]:
# "NAPS ID" means Station
station_counts = ontario_NO2_2013.groupby("City")["NAPS ID"].nunique()
print(station_counts)

City
Barrie              1
Belleville          1
Brampton            1
Brantford           1
Burlington          1
Chatham             1
Cornwall            1
Grand Bend          1
Guelph              1
Hamilton            3
Kingston            2
Kitchener           1
London              1
Mississauga         1
Newmarket           1
North Bay           1
Oakville            1
Oshawa              1
Ottawa              2
Parry Sound         1
Peterborough        1
Sarnia              1
Sault Ste. Marie    1
Simcoe              1
St. Catharines      1
Sudbury             1
Thunder Bay         1
Tiverton            1
Toronto             5
Windsor             2
Name: NAPS ID, dtype: int64


From the results, we can confirm that the duplicates come from mutiple stations. For example, in Toronto, there are 5 stations, so Totonto has 1825 values in 2013


### Data Type

In [71]:
#Convert 'Data' from object to Datetime

ontario_NO2_2013["Date"] = pd.to_datetime(ontario_NO2_2013["Date"], format="%Y-%m-%d", errors="raise")
print(ontario_NO2_2013.info())

<class 'pandas.core.frame.DataFrame'>
Index: 14235 entries, 13505 to 27739
Data columns (total 31 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           14235 non-null  object        
 1   NAPS ID             14235 non-null  int64         
 2   City                14235 non-null  object        
 3   Province/Territory  14235 non-null  object        
 4   Latitude            14235 non-null  float64       
 5   Longitude           14235 non-null  float64       
 6   Date                14235 non-null  datetime64[ns]
 7   H01                 14235 non-null  int64         
 8   H02                 14235 non-null  int64         
 9   H03                 14235 non-null  int64         
 10  H04                 14235 non-null  int64         
 11  H05                 14235 non-null  int64         
 12  H06                 14235 non-null  int64         
 13  H07                 14235 non-null  int64      

### Calculate 3h Max for each day (row) and each station

In [72]:
# 1. Define the hourly columns H01 through H24
hour_cols = [f'H{i:02d}' for i in range(1, 25)]

# 2. Replace invalid values (-999) with NaN so they are excluded from calculations
ontario_NO2_2013[hour_cols] = ontario_NO2_2013[hour_cols].replace({-999: np.nan})

# 3. Compute the 3-hour rolling average for each row, then take the maximum of those averages
ontario_NO2_2013['NO2_3hMax'] = (
    ontario_NO2_2013[hour_cols]
    .apply(lambda row: row.rolling(window=3, min_periods=3).mean().max(), axis=1)
).round(2)

# 4. Print the first few rows to verify the new column
print(ontario_NO2_2013[['Date','NO2_3hMax']].head())

            Date  NO2_3hMax
13505 2013-01-01      27.67
13506 2013-01-02      33.67
13507 2013-01-03      35.00
13508 2013-01-04      21.67
13509 2013-01-05        NaN


In [73]:
# Count each unique value and show the result
counts_3hMax = ontario_NO2_2013["NO2_3hMax"].value_counts().sort_index()

# Display the result
print(counts_3hMax)

NO2_3hMax
0.00      8
0.33      6
0.67      6
1.00     67
1.33     51
         ..
65.67     1
66.33     1
71.33     1
75.00     1
77.00     1
Name: count, Length: 182, dtype: int64


In [74]:
ontario_NO2_2013.info()


<class 'pandas.core.frame.DataFrame'>
Index: 14235 entries, 13505 to 27739
Data columns (total 32 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           14235 non-null  object        
 1   NAPS ID             14235 non-null  int64         
 2   City                14235 non-null  object        
 3   Province/Territory  14235 non-null  object        
 4   Latitude            14235 non-null  float64       
 5   Longitude           14235 non-null  float64       
 6   Date                14235 non-null  datetime64[ns]
 7   H01                 13704 non-null  float64       
 8   H02                 13694 non-null  float64       
 9   H03                 13701 non-null  float64       
 10  H04                 13270 non-null  float64       
 11  H05                 13651 non-null  float64       
 12  H06                 13696 non-null  float64       
 13  H07                 13687 non-null  float64    

### Calculate 8h Max

In [76]:
# 1. Define the hourly columns H01 through H24
hour_cols = [f'H{i:02d}' for i in range(1, 25)]

# 2. Replace invalid values (-999) with NaN so they are excluded from calculations
ontario_NO2_2013[hour_cols] = ontario_NO2_2013[hour_cols].replace({-999: np.nan})

# 3. Compute the 8-hour rolling average for each row, then take the maximum of those averages
ontario_NO2_2013['NO2_8hMax'] = (
    ontario_NO2_2013[hour_cols]
    .apply(lambda row: row.rolling(window=8, min_periods=8).mean().max(), axis=1)
).round(2)

# 4. Print the first few rows to verify the new column
print(ontario_NO2_2013[['Date','NO2_8hMax']].head())

            Date  NO2_8hMax
13505 2013-01-01      19.75
13506 2013-01-02      26.62
13507 2013-01-03      28.38
13508 2013-01-04      17.50
13509 2013-01-05        NaN


In [77]:
# Count each unique value and show the result
counts_8hMax = ontario_NO2_2013["NO2_8hMax"].value_counts().sort_index()

# Display the result
print(counts_8hMax)

NO2_8hMax
0.00     8
0.12     5
0.25     4
0.38     6
0.50     8
        ..
52.25    1
53.50    2
60.88    1
62.62    1
74.50    1
Name: count, Length: 375, dtype: int64


In [78]:
ontario_NO2_2013.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14235 entries, 13505 to 27739
Data columns (total 33 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           14235 non-null  object        
 1   NAPS ID             14235 non-null  int64         
 2   City                14235 non-null  object        
 3   Province/Territory  14235 non-null  object        
 4   Latitude            14235 non-null  float64       
 5   Longitude           14235 non-null  float64       
 6   Date                14235 non-null  datetime64[ns]
 7   H01                 13704 non-null  float64       
 8   H02                 13694 non-null  float64       
 9   H03                 13701 non-null  float64       
 10  H04                 13270 non-null  float64       
 11  H05                 13651 non-null  float64       
 12  H06                 13696 non-null  float64       
 13  H07                 13687 non-null  float64    

### Annual Max 3-hour/8-hour Average NO2 Concentration for each Station (NAPS ID )/City

Since we want to calculate annual mean, so we don't need to fill the missing values for 3hMax and 8hMax, unless no data in a station


In [83]:
# 1. Extract the year from the Date column
ontario_NO2_2013['Year'] = pd.to_datetime(ontario_NO2_2013['Date']).dt.year

# 2. Group by Year, City, and NAPS ID to compute annual mean, 
#    and carry through station metadata
NO2_annual_mean_2013 = (
    ontario_NO2_2013
    .groupby(['Year', 'City', 'NAPS ID'], as_index=False)
    .agg(
        Pollutant    = ('Pollutant',    'first'),
        Latitude     = ('Latitude',     'first'),
        Longitude    = ('Longitude',    'first'),
        NO2_annu_mean_3hMax = ('NO2_3hMax',       'mean'),           
        NO2_annu_mean_8hMax = ('NO2_8hMax',       'mean') )           
)

# 3. Sort for readability
NO2_annual_mean_2013.sort_values(['Year', 'City', 'NAPS ID'], inplace=True)

# 4. Inspect the first few rows
print(NO2_annual_mean_2013.head())

# 5. Check the structure of the result
NO2_annual_mean_2013.info()


   Year        City  NAPS ID Pollutant  Latitude  Longitude  \
0  2013      Barrie    65001       NO2  44.38231  -79.70243   
1  2013  Belleville    65401       NO2  44.15053  -77.39550   
2  2013    Brampton    60428       NO2  43.69875  -79.78092   
3  2013   Brantford    61402       NO2  43.13861  -80.29264   
4  2013  Burlington    63001       NO2  43.31457  -79.80276   

   NO2_annu_mean_3hMax  NO2_annu_mean_8hMax  
0            15.432055            11.722582  
1            10.358159             7.399038  
2            17.492473            13.510797  
3             9.155178             7.058630  
4            20.432473            15.854286  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39 entries, 0 to 38
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Year                 39 non-null     int32  
 1   City                 39 non-null     object 
 2   NAPS ID              39 non-null     i

Check missing values for annual_max_3h and annual_max_8h.

In [85]:
# 1. Which stations got zero valid daily 3hMax values?
zero_3h = (
    ontario_NO2_2013
    .groupby('NAPS ID')['NO2_3hMax']
    .apply(lambda s: s.notna().sum())
)
print("Stations with no valid 3hMax days:")
print(zero_3h[zero_3h == 0])


# 2. Same for 8-hour
zero_8h = (
    ontario_NO2_2013
    .groupby('NAPS ID')['NO2_8hMax']
    .apply(lambda s: s.notna().sum())
)
print("\nStations with no valid 8hMax days:")
print(zero_8h[zero_8h == 0])

# 3. Inspect one of those stations’ daily data to see why
bad_id = zero_3h[zero_3h == 0].index[0]
print(f"\nDaily 3hMax for station {bad_id}:")
print(ontario_NO2_2013.loc[ontario_NO2_2013['NAPS ID']==bad_id, ['Date','NO2_3hMax']].head(10))


Stations with no valid 3hMax days:
NAPS ID
60304    0
Name: NO2_3hMax, dtype: int64

Stations with no valid 8hMax days:
NAPS ID
60304    0
Name: NO2_8hMax, dtype: int64

Daily 3hMax for station 60304:
            Date  NO2_3hMax
15330 2013-01-01        NaN
15331 2013-01-02        NaN
15332 2013-01-03        NaN
15333 2013-01-04        NaN
15334 2013-01-05        NaN
15335 2013-01-06        NaN
15336 2013-01-07        NaN
15337 2013-01-08        NaN
15338 2013-01-09        NaN
15339 2013-01-10        NaN


From the result, we can see that the station '60304' in Kingston never produced any valid data in 2013, which is same with PM2.5 in 2013. Need to check other years to see how to cope this issue


# 4. Save out to CSV 
out_path = r"air_pollution_data\NO2_2013_ontario_annual_by_station.csv"
NO2_annual_mean_2013.to_csv(out_path, index=False)
print(f"Saved annual station stats to {out_path}")

# O3 from 2013 to 2022

## O3 in 2013

In [86]:
# Read the 2013 PM2.5 file, skipping the first 7 lines of metadata
O3_2013 = pd.read_csv(
    r"air_pollution_data\O3_2013.csv",
    skiprows=7,      # Skip lines 1–7 (6 description lines + 1 blank line)
    header=0,        # Use line 8 as the column names
    nrows=78475,     # Read the remaining 78475 data rows (from line 9 to line 78,483)
    engine="python"  # Ensure skiprows + nrows are handled properly
)

# Keep only the English part of each column name (text before "//") and strip whitespace
O3_2013.columns = [col.split("//")[0].strip() for col in O3_2013.columns]

#Filter for Ontario
ontario_O3_2013 = O3_2013[O3_2013["Province/Territory"] == "ON"].copy()

# Verify the cleaned column names and dataset shape
print(O3_2013.columns.tolist())

print(f"All rows: {O3_2013.shape[0]}, Ontario rows: {ontario_O3_2013.shape[0]}")



['Pollutant', 'NAPS ID', 'City', 'Province/Territory', 'Latitude', 'Longitude', 'Date', 'H01', 'H02', 'H03', 'H04', 'H05', 'H06', 'H07', 'H08', 'H09', 'H10', 'H11', 'H12', 'H13', 'H14', 'H15', 'H16', 'H17', 'H18', 'H19', 'H20', 'H21', 'H22', 'H23', 'H24']
All rows: 78475, Ontario rows: 17520


In [91]:
# Display the first 6 rows in a DataFrame format
O3_2013.head(6)



Pollutant  NAPS ID        City Province/Territory  Latitude  Longitude  \
0        O3    10102  St. John's                 NL  47.56038  -52.71147   
1        O3    10102  St. John's                 NL  47.56038  -52.71147   
2        O3    10102  St. John's                 NL  47.56038  -52.71147   
3        O3    10102  St. John's                 NL  47.56038  -52.71147   
4        O3    10102  St. John's                 NL  47.56038  -52.71147   
5        O3    10102  St. John's                 NL  47.56038  -52.71147   

         Date  H01  H02  H03  ...  H15  H16  H17  H18  H19  H20  H21  H22  \
0  2013-01-01   37   37   39  ...   40   40   39   35   33   33   31   35   
1  2013-01-02   32   33   32  ...   33   31   25   28   24   22   21   20   
2  2013-01-03   32   31   30  ...   30   29   30   27   24   32   31   33   
3  2013-01-04   34   35   35  ...   34   32   33   34   34   34   35   36   
4  2013-01-05   34   32   32  ...   31   31   30   32   33   32   30   31   
5  2013-01-06   33   35   35  ...   35   33   31   30   30   31   30   29   

   H23  H24  
0   35   35  
1   26   29  
2   34   33  
3   34   35  
4   33   33  
5   31   32  

[6 rows x 31 columns]

In [92]:
O3_2013.tail(6)

Pollutant  NAPS ID          City Province/Territory  Latitude  \
78469        O3   129501  Snare Rapids                 NT  63.50825   
78470        O3   129501  Snare Rapids                 NT  63.50825   
78471        O3   129501  Snare Rapids                 NT  63.50825   
78472        O3   129501  Snare Rapids                 NT  63.50825   
78473        O3   129501  Snare Rapids                 NT  63.50825   
78474        O3   129501  Snare Rapids                 NT  63.50825   

       Longitude        Date  H01  H02  H03  ...  H15  H16  H17  H18  H19  \
78469 -116.00867  2013-12-26   21   24   23  ...   27   27   27   27   27   
78470 -116.00867  2013-12-27   26   26   26  ...   26   26   26   26   26   
78471 -116.00867  2013-12-28   27   27   27  ...   28   28   28   28   27   
78472 -116.00867  2013-12-29   25   24   23  ...   22 -999 -999 -999 -999   
78473 -116.00867  2013-12-30   25   25   25  ...   19   18   20   19   17   
78474 -116.00867  2013-12-31   16   16   15  ...   16   16   16   15   14   

       H20  H21  H22  H23  H24  
78469   27   26   26   26   26  
78470   26   26   26   27   27  
78471   27   27   27   26   25  
78472 -999   24   25   25   25  
78473   18   17   16   16   16  
78474   16   15   14   14 -999  

[6 rows x 31 columns]

In [89]:
# Display the first 6 rows in a DataFrame format
ontario_O3_2013.head(6)


Pollutant  NAPS ID    City Province/Territory  Latitude  Longitude  \
28835        O3    60104  Ottawa                 ON  45.43433    -75.676   
28836        O3    60104  Ottawa                 ON  45.43433    -75.676   
28837        O3    60104  Ottawa                 ON  45.43433    -75.676   
28838        O3    60104  Ottawa                 ON  45.43433    -75.676   
28839        O3    60104  Ottawa                 ON  45.43433    -75.676   
28840        O3    60104  Ottawa                 ON  45.43433    -75.676   

             Date  H01  H02  H03  ...  H15  H16  H17  H18  H19  H20  H21  H22  \
28835  2013-01-01   28   29   28  ...   34   34   32   28   22   23   11   13   
28836  2013-01-02    2    6   10  ...   35   33   30   26   22   21   22   14   
28837  2013-01-03   24   24   20  ... -999 -999 -999 -999 -999 -999 -999 -999   
28838  2013-01-04 -999 -999 -999  ...   20   14   17   20   23   24   25   27   
28839  2013-01-05   33   37   37  ...   30   29   23   11    4   10    5    3   
28840  2013-01-06   15   23   23  ...   21   22   18   16   20   17   22   24   

       H23  H24  
28835   15    9  
28836   16   20  
28837 -999 -999  
28838   27   29  
28839    4   10  
28840   26   29  

[6 rows x 31 columns]

In [97]:
#Display info
ontario_O3_2013.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17520 entries, 28835 to 46354
Data columns (total 31 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Pollutant           17520 non-null  object 
 1   NAPS ID             17520 non-null  int64  
 2   City                17520 non-null  object 
 3   Province/Territory  17520 non-null  object 
 4   Latitude            17520 non-null  float64
 5   Longitude           17520 non-null  float64
 6   Date                17520 non-null  object 
 7   H01                 17520 non-null  int64  
 8   H02                 17520 non-null  int64  
 9   H03                 17520 non-null  int64  
 10  H04                 17520 non-null  int64  
 11  H05                 17520 non-null  int64  
 12  H06                 17520 non-null  int64  
 13  H07                 17520 non-null  int64  
 14  H08                 17520 non-null  int64  
 15  H09                 17520 non-null  int64  
 16  H10  

In [98]:
# Unique values in City

# 1. Number of unique cities
num_cities = ontario_O3_2013["City"].nunique(dropna=True)
print(f"Total unique cities: {num_cities}")

# 2. List of unique city names
unique_cities = ontario_O3_2013["City"].dropna().unique()
print("Unique city names:")
print(unique_cities)

# 3. How many readings per city
city_counts = ontario_O3_2013["City"].value_counts()
print("\nObservations per city:")
print(city_counts)

Total unique cities: 39
Unique city names:
['Ottawa' 'Windsor' 'Kingston' 'Toronto' 'Brampton' 'Mississauga'
 'Hamilton' 'Sudbury' 'Sault Ste. Marie' 'Thunder Bay' 'London' 'Sarnia'
 'Peterborough' 'Cornwall' 'St. Catharines' 'Brantford' 'Kitchener'
 'Oakville' 'Oshawa' 'Guelph' 'North Bay' 'Tiverton' 'Simcoe' 'Burlington'
 'Dorset' 'Grand Bend' 'Exp. Lakes Area' 'Algoma' 'Egbert' 'Barrie'
 'Newmarket' 'Parry Sound' 'Port Stanley' 'Belleville' 'Morrisburg'
 'Chatham' 'Pickle Lake' 'Moonbeam' 'Petawawa']

Observations per city:
City
Toronto             1825
Hamilton            1095
Ottawa               730
Kingston             730
Windsor              730
Brampton             365
Mississauga          365
Sudbury              365
Sault Ste. Marie     365
Thunder Bay          365
London               365
Sarnia               365
Peterborough         365
Cornwall             365
St. Catharines       365
Brantford            365
Kitchener            365
Oakville             365
Oshawa      

From the results we can see that there are duplicates in some cities, like Toronto, Hamilton, Ottawa


In [99]:
# "NAPS ID" means Station
station_counts = ontario_O3_2013.groupby("City")["NAPS ID"].nunique()
print(station_counts)


City
Algoma              1
Barrie              1
Belleville          1
Brampton            1
Brantford           1
Burlington          1
Chatham             1
Cornwall            1
Dorset              1
Egbert              1
Exp. Lakes Area     1
Grand Bend          1
Guelph              1
Hamilton            3
Kingston            2
Kitchener           1
London              1
Mississauga         1
Moonbeam            1
Morrisburg          1
Newmarket           1
North Bay           1
Oakville            1
Oshawa              1
Ottawa              2
Parry Sound         1
Petawawa            1
Peterborough        1
Pickle Lake         1
Port Stanley        1
Sarnia              1
Sault Ste. Marie    1
Simcoe              1
St. Catharines      1
Sudbury             1
Thunder Bay         1
Tiverton            1
Toronto             5
Windsor             2
Name: NAPS ID, dtype: int64


From the results, we can confirm that the duplicates come from mutiple stations. For example, in Toronto, there are 5 stations, so Totonto has 1825 values in 2013


### Data Type

In [ ]:
# Convert 'Data' from object to Datetime
ontario_O3_2013["Date"] = pd.to_datetime(ontario_O3_2013["Date"], format="%Y-%m-%d", errors="raise")
print(ontario_O3_2013.info())

<class 'pandas.core.frame.DataFrame'>
Index: 17520 entries, 28835 to 46354
Data columns (total 31 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           17520 non-null  object        
 1   NAPS ID             17520 non-null  int64         
 2   City                17520 non-null  object        
 3   Province/Territory  17520 non-null  object        
 4   Latitude            17520 non-null  float64       
 5   Longitude           17520 non-null  float64       
 6   Date                17520 non-null  datetime64[ns]
 7   H01                 17520 non-null  int64         
 8   H02                 17520 non-null  int64         
 9   H03                 17520 non-null  int64         
 10  H04                 17520 non-null  int64         
 11  H05                 17520 non-null  int64         
 12  H06                 17520 non-null  int64         
 13  H07                 17520 non-null  int64      

### Calculate 3h Max for each day (row) and each station

In [101]:
# 1. Define the hourly columns H01 through H24
hour_cols = [f'H{i:02d}' for i in range(1, 25)]

# 2. Replace invalid values (-999) with NaN so they are excluded from calculations
ontario_O3_2013[hour_cols] = ontario_O3_2013[hour_cols].replace({-999: np.nan})

# 3. Compute the 3-hour rolling average for each row, then take the maximum of those averages
ontario_O3_2013['O3_3hMax'] = (
    ontario_O3_2013[hour_cols]
    .apply(lambda row: row.rolling(window=3, min_periods=3).mean().max(), axis=1)
).round(2)

# 4. Print the first few rows to verify the new column
print(ontario_O3_2013[['Date','O3_3hMax']].head())

            Date  O3_3hMax
28835 2013-01-01     33.33
28836 2013-01-02     34.67
28837 2013-01-03     22.67
28838 2013-01-04     27.67
28839 2013-01-05     36.67


In [102]:
# Count each unique value and show the result
counts_3hMax = ontario_O3_2013["O3_3hMax"].value_counts().sort_index()

# Display the result
print(counts_3hMax)

O3_3hMax
4.67     2
5.00     1
5.67     1
6.33     1
6.67     1
        ..
87.67    1
88.67    1
89.67    1
91.67    1
95.00    1
Name: count, Length: 240, dtype: int64


In [103]:
ontario_O3_2013.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17520 entries, 28835 to 46354
Data columns (total 32 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           17520 non-null  object        
 1   NAPS ID             17520 non-null  int64         
 2   City                17520 non-null  object        
 3   Province/Territory  17520 non-null  object        
 4   Latitude            17520 non-null  float64       
 5   Longitude           17520 non-null  float64       
 6   Date                17520 non-null  datetime64[ns]
 7   H01                 16654 non-null  float64       
 8   H02                 16640 non-null  float64       
 9   H03                 16647 non-null  float64       
 10  H04                 16196 non-null  float64       
 11  H05                 16438 non-null  float64       
 12  H06                 16594 non-null  float64       
 13  H07                 16641 non-null  float64    

### Calculate 8h Max

In [104]:
# 1. Define the hourly columns H01 through H24
hour_cols = [f'H{i:02d}' for i in range(1, 25)]

# 2. Replace invalid values (-999) with NaN so they are excluded from calculations
ontario_O3_2013[hour_cols] = ontario_O3_2013[hour_cols].replace({-999: np.nan})

# 3. Compute the 8-hour rolling average for each row, then take the maximum of those averages
ontario_O3_2013['O3_8hMax'] = (
    ontario_O3_2013[hour_cols]
    .apply(lambda row: row.rolling(window=8, min_periods=8).mean().max(), axis=1)
).round(2)

# 4. Print the first few rows to verify the new column
print(ontario_O3_2013[['Date','O3_8hMax']].head())

            Date  O3_8hMax
28835 2013-01-01     31.00
28836 2013-01-02     33.88
28837 2013-01-03     18.00
28838 2013-01-04     24.00
28839 2013-01-05     34.00


In [105]:
# Count each unique value and show the result
counts_8hMax = ontario_O3_2013["O3_8hMax"].value_counts().sort_index()

# Display the result
print(counts_8hMax)

O3_8hMax
3.50     1
3.88     2
4.12     1
4.38     2
4.75     2
        ..
79.75    2
81.12    1
81.25    2
81.38    1
81.50    1
Name: count, Length: 551, dtype: int64


In [106]:
ontario_O3_2013.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17520 entries, 28835 to 46354
Data columns (total 33 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Pollutant           17520 non-null  object        
 1   NAPS ID             17520 non-null  int64         
 2   City                17520 non-null  object        
 3   Province/Territory  17520 non-null  object        
 4   Latitude            17520 non-null  float64       
 5   Longitude           17520 non-null  float64       
 6   Date                17520 non-null  datetime64[ns]
 7   H01                 16654 non-null  float64       
 8   H02                 16640 non-null  float64       
 9   H03                 16647 non-null  float64       
 10  H04                 16196 non-null  float64       
 11  H05                 16438 non-null  float64       
 12  H06                 16594 non-null  float64       
 13  H07                 16641 non-null  float64    

### Annual Max 3-hour/8-hour Average O3 Concentration for each Station (NAPS ID )/City

Since we want to calculate annual mean, so we don't need to fill the missing values for 3hMax and 8hMax, unless no data in a station

In [ ]:
# 1. Extract the year from the Date column
ontario_O3_2013['Year'] = pd.to_datetime(ontario_O3_2013['Date']).dt.year

# 2. Group by Year, City, and NAPS ID to compute annual mean, 
#    and carry through station metadata
O3_annual_mean_2013 = (
    ontario_O3_2013
    .groupby(['Year', 'City', 'NAPS ID'], as_index=False)
    .agg(
        Pollutant    = ('Pollutant',    'first'),
        Latitude     = ('Latitude',     'first'),
        Longitude    = ('Longitude',    'first'),
        O3_annu_mean_3hMax = ('O3_3hMax',       'mean'),           
        O3_annu_mean_8hMax = ('O3_8hMax',       'mean') )           
)

# 3. Sort for readability
O3_annual_mean_2013.sort_values(['Year', 'City', 'NAPS ID'], inplace=True)

# 4. Inspect the first few rows
print(O3_annual_mean_2013.head())

# 5. Check the structure of the result
O3_annual_mean_2013.info()

   Year        City  NAPS ID Pollutant  Latitude  Longitude  \
0  2013      Algoma    64101        O3  47.03361  -84.37889   
1  2013      Barrie    65001        O3  44.38231  -79.70243   
2  2013  Belleville    65401        O3  44.15053  -77.39550   
3  2013    Brampton    60428        O3  43.69875  -79.78092   
4  2013   Brantford    61402        O3  43.13861  -80.29264   

   O3_annu_mean_3hMax  O3_annu_mean_8hMax  
0           38.948932           37.127967  
1           35.113205           32.747802  
2           39.215055           36.769337  
3           38.653945           35.625616  
4           41.105068           38.450877  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Year                48 non-null     int32  
 1   City                48 non-null     object 
 2   NAPS ID             48 non-null     int64  
 3   Pollu

In [109]:
print(O3_annual_mean_2013)

    Year              City  NAPS ID Pollutant  Latitude  Longitude  \
0   2013            Algoma    64101        O3  47.03361  -84.37889   
1   2013            Barrie    65001        O3  44.38231  -79.70243   
2   2013        Belleville    65401        O3  44.15053  -77.39550   
3   2013          Brampton    60428        O3  43.69875  -79.78092   
4   2013         Brantford    61402        O3  43.13861  -80.29264   
5   2013        Burlington    63001        O3  43.31457  -79.80276   
6   2013           Chatham    65801        O3  42.40369  -82.20831   
7   2013          Cornwall    61201        O3  45.01798  -74.73531   
8   2013            Dorset    63301        O3  45.22428  -78.93294   
9   2013            Egbert    64401        O3  44.23106  -79.78310   
10  2013   Exp. Lakes Area    64001        O3  49.66389  -93.72111   
11  2013        Grand Bend    63701        O3  43.33319  -81.74282   
12  2013            Guelph    61802        O3  43.55164  -80.26416   
13  2013          Ha

Check missing values for annual_max_3h and annual_max_8h

In [108]:
# 1. Which stations got zero valid daily 3hMax values?
zero_3h = (
    ontario_O3_2013
    .groupby('NAPS ID')['O3_3hMax']
    .apply(lambda s: s.notna().sum())
)
print("Stations with no valid 3hMax days:")
print(zero_3h[zero_3h == 0])


# 2. Same for 8-hour
zero_8h = (
    ontario_O3_2013
    .groupby('NAPS ID')['O3_8hMax']
    .apply(lambda s: s.notna().sum())
)
print("\nStations with no valid 8hMax days:")
print(zero_8h[zero_8h == 0])

# 3. Inspect one of those stations’ daily data to see why
bad_id = zero_3h[zero_3h == 0].index[0]
print(f"\nDaily 3hMax for station {bad_id}:")
print(ontario_O3_2013.loc[ontario_O3_2013['NAPS ID']==bad_id, ['Date','O3_3hMax']].head(10))


Stations with no valid 3hMax days:
NAPS ID
60304    0
65901    0
Name: O3_3hMax, dtype: int64

Stations with no valid 8hMax days:
NAPS ID
60304    0
65901    0
Name: O3_8hMax, dtype: int64

Daily 3hMax for station 60304:
            Date  O3_3hMax
30660 2013-01-01       NaN
30661 2013-01-02       NaN
30662 2013-01-03       NaN
30663 2013-01-04       NaN
30664 2013-01-05       NaN
30665 2013-01-06       NaN
30666 2013-01-07       NaN
30667 2013-01-08       NaN
30668 2013-01-09       NaN
30669 2013-01-10       NaN


The station '60304' in Kingston and '65901' in Pickle Lake never produced any valid data in 2013, need to check other years to see how to cope this issue.

Note: The station '60304' in Kingston never produced any valid data of PM2.5, NO2, and O3 in 2013


# 4. Save out to CSV 
out_path = r"air_pollution_data\O3_2013_ontario_annual_by_station.csv"
O3_annual_mean_2013.to_csv(out_path, index=False)
print(f"Saved annual station stats to {out_path}")